src/Product/Importer.php

```diff
@@ -220,15 +222,29 @@ private function handleChannels(ProductInterface $product, array $productVariant
 
     private function handleTaxons(ProductInterface $product, array $akeneoProduct): void
     {
-        $taxons = $this->taxonsResolver->resolve($akeneoProduct);
-        foreach ($taxons as $taxon) {
-            if ($product->hasTaxon($taxon)) {
-                continue;
+        $akeneoTaxons = new ArrayCollection($this->taxonsResolver->resolve($akeneoProduct));
+        $syliusTaxons = $product->getTaxons();
+        $akeneoTaxonsCodes = $akeneoTaxons->map(function (TaxonInterface $taxon) {return $taxon->getCode(); })->toArray();
+        $syliusTaxonsCodes = $syliusTaxons->map(function (TaxonInterface $taxon) {return $taxon->getCode(); })->toArray();
+        $toAddTaxons = $akeneoTaxons->filter(
+            function (TaxonInterface $taxon) use ($syliusTaxonsCodes) {
+                return !in_array($taxon->getCode(), $syliusTaxonsCodes, true);
             }
-            $productTaxon = $this->productTaxonFactory->createNew();
-            Assert::isInstanceOf($productTaxon, ProductTaxonInterface::class);
+        );
+        $toRemoveTaxonsCodes = array_diff($syliusTaxonsCodes, $akeneoTaxonsCodes);
+
+        foreach ($product->getProductTaxons() as $productTaxon) {
+            $taxon = $productTaxon->getTaxon();
+            Assert::isInstanceOf($taxon, TaxonInterface::class);
+            if (in_array($taxon->getCode(), $toRemoveTaxonsCodes, true)) {
+                $product->getProductTaxons()->removeElement($productTaxon);
+            }
+        }
+        foreach ($toAddTaxons as $toAddTaxon) {
```

Comment:
To be more uniform with the previous loop, I would remove the call to filter() at line 229 and replace it with another call to array_diff():

$toAddTaxonsCodes = array_diff($akeneoTaxonsCodes, $syliusTaxonsCodes);

And then here I would change this foreach to iterate on $akeneoTaxons and check with in_array() if the current taxon must be added

Reply:
This was my first implementation but when adding taxons we need the actual Taxon objects because we have to set them to the ProductTaxon objects. Having only the taxon code force to do another search to get the taxon object.

Reply:
I think it's a matter of preference. I'd prefer to avoid the check on which are the taxons to add instead of having uniform array_diff.